src/auto-evo/Miche.cs

```diff
@@ -0,0 +1,210 @@
+﻿namespace AutoEvo;
+
+using System.Collections.Generic;
+using System.Linq;
+
+/// <summary>
+///   A node for the Miche Tree
+/// </summary>
+/// <remarks>
+///   <para>
+///     The Miche class forms a tree by storing a list of child instances of Miche Nodes. If a Miche has no children
+///     it is considered a leaf node and can have a species Occupant instead. This class handles insertion into the
+///     tree through scores from the selection pressure it contains. For a fuller explanation see docs/auto_evo.md
+///   </para>
+/// </remarks>
+public class Miche
+{
+    public Miche? Parent;
+    public List<Miche> Children = new();
+
+    // Occupant should always be null if this Miche has children
+    public Species? Occupant;
+    public SelectionPressure Pressure;
+
+    public Miche(SelectionPressure pressure)
+    {
+        Pressure = pressure;
+    }
+
+    public bool IsLeafNode()
+    {
+        return Children.Count == 0;
+    }
+
+    public Miche Root()
+    {
+        if (Parent == null)
+        {
+            return this;
+        }
+
+        return Parent.Root();
+    }
+
+    public IEnumerable<Miche> GetLeafNodes()
+    {
+        return GetLeafNodes(new List<Miche>());
+    }
+
+    public IEnumerable<Miche> GetLeafNodes(List<Miche> nodes)
+    {
+        if (IsLeafNode())
+        {
+            nodes.Add(this);
+            return nodes;
+        }
+
+        foreach (var child in Children)
+        {
+            nodes.AddRange(child.GetLeafNodes());
+        }
+
+        return nodes;
+    }
+
+    public IEnumerable<Species> GetOccupants()
+    {
+        return GetOccupants(new List<Species>());
+    }
+
+    public IEnumerable<Species> GetOccupants(List<Species> occupants)
+    {
+        if (Occupant != null)
+        {
+            occupants.Add(Occupant);
+            return occupants;
+        }
+
+        foreach (var child in Children)
+        {
+            occupants.AddRange(child.GetOccupants());
+        }
+
+        return occupants;
+    }
+
+    public IEnumerable<Miche> BackTraversal()
+    {
+        return BackTraversal(new List<Miche>());
+    }
+
+    public IEnumerable<Miche> BackTraversal(List<Miche> currentTraversal)
+    {
+        currentTraversal.Insert(0, this);
+
+        if (Parent == null)
+            return currentTraversal;
+
+        return Parent.BackTraversal(currentTraversal);
+    }
+
+    public void AddChild(Miche newChild)
+    {
+        Children.Add(newChild);
+        newChild.Parent = this;
+    }
+
+    public void AddChildren(IEnumerable<Miche> newChildren)
+    {
+        foreach (var child in newChildren)
+        {
+            AddChild(child);
+        }
+    }
+
+    public bool InsertSpecies(Species species, SimulationCache cache, bool dry = false)
+    {
+        var scoresDictionary = GetOccupants().Distinct().ToDictionary(x => x, _ => 0.0f);
```

Comment:
I think avoiding LINQ here would be good as this probably allocates memory.

Also I saw that Distinct is used quite many times so it would probably be pretty good to create a separate GetDistinctOccupants method (that takes in a HashSet<Species> as the parameter to fill as the result).

Reply:
I added a GetDistinctOccupants, but I quickly realized that there was not a single place in the code that I wanted to use the original GetOccupants. As any usecase is pretty wall covered by GetLeafNodes, I just made GetDistinctOccupants into GetOccupants and replaced it entirely.